enhancements/substitutes-for.md

```diff
@@ -0,0 +1,430 @@
+consumed by the CVE rebuilder.  So substitutes-for will need to be a GA api with
+a GA implementation in OPM.
+
+### Upgrade / Downgrade Strategy
+
+All logic related to this behavior will be tied to the opm binary used to
+generate the catalog, and the opm binary in the catalog image that is used
+to serve the content.  Therefore it should not require any changes to 
+OLM on the target cluster.  Catalogs produced that use this feature must
+work with any existing supported OCP cluster.
+
+Users who want to take advantage of this feature will need to use a current
+opm binary both to build their catalog DB as well as packaged in their 
+catalog image.
+
+Note:  Today the operator pipeline uses the latest opm binary to build catalog DBs
+(good) but matches the opm binary in the catalog to the ocp version.  We will
+need to backport the substitutes-for support in opm to all opm versions needed for
+the operator pipeline (today that means back to 4.5 at least).
+
+If an existing catalog DB is being modified by this new opm binary, it will need
+to add the new table/column tracking the substitutes-for metadata.
+
+
+### Version Skew Strategy
+
+See upgrade/downgrade strategy for discussion of these issues.
+
+
+## Drawbacks
+
+This is bolting on yet another piece of CSV metadata that influences the upgrade graph.  The upgrade
+graph (when using skips/replaces/skiprange) is already often misunderstood by operator authors
+and this adds another piece to reason about.  
+
+Furthermore, the actual replaces/skips metadata can't be seen by just looking at the bundle itself since 
+the references are mutated only in the DB based on what substitutes-for relationships exist, making it even 
+harder to validate the upgrade graph reflects what is desired.  Unfortunately that is a necessary 
+consequence of the fact that we are trying to achieve a way to modify what bundles an already shipped 
+bundle can upgrade from, without re-shipping the bundle.  So post-publication mutation is a requirement.
+
+
+
+## Alternatives
+
+### Mutate in ListBundles
+
+Rather than modifying the metadata in the catalog DB as bundles are added, we could mutate
+the content as it is served by ListBundles.  This has the advantage of only processing the
+substitution relationships when they are definitely needed and the complete set is known,
+however it also means processing that logic every single time the api is invoked, which is
+ultimately more costly than doing it once on each bundle add.
+
+### Declarative Config
+
+We can abandon skips/replaces entirely and moving to an explicitly defined upgrade graph.  
+The CVE rebuilder could then stitch new content into that upgrade graph as it is produced.
+The primary arguments against this approach, for now, are:
+
+1) we need something sooner than that will be ready
+2) even if we had it, using it would require operator teams to migrate to the declarative config
```

Comment:
Potentially, operator teams would not need to migrate at all, and freshmaker would be the only thing to directly modify the declarative config (i.e. by maintaining a list of "substitutes-for" with a tool that can apply the appropriate transformations to the declarative config). (basing this on the current discussions of a low-level declarative config and a high-level user input)